### PR TITLE
Try to break resolve into more isolated parts

### LINKED ISSUE #63223: Compiler panic on a macro invocation combined with invalid pub declaration
The compiler panics when compiling the following code. I had been expecting to see syntax errors instead. Deleting either line causes the compiler not to panic and give expected error messages.

```rust
const FOO: &str = b!("foo");
 
pub struct Bar { pub(in ::a) x: u64 }

```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c43d1f37cad49bb52a9ab01461792b11))

### PR DESCRIPTION
**Pull request by petrochenkov**
Some small step towards resolve librarification.

"Late resolution" is the pass that resolves most of names in a crate beside imports and macros.
It runs when the crate is fully expanded and its module structure is fully built.
So we just walk through the crate and resolve all the expressions, types, etc.

This pass is pretty self-contained, but it was previously done by implementing `Visitor` on the whole `Resolver` (which is used for many other tasks), and fields specific to this pass were indiscernible from the global `Resolver` state.

This PR moves the late resolution pass into a separate visitor and a separate file, fields specific to this visitor are moved from `Resolver` as well.

I'm especially happy about `current_module` being removed from `Resolver`.
It was used even for operations not related to visiting and changing the `current_module` position in process.
It was also used as an implicit argument for some functions used in this style
```rust
let orig_current_module = mem::replace(&mut self.current_module, module);
self.resolve_ident_somewhere();
self.current_module = orig_current_module;
```
and having effects on e.g. privacy checking somewhere deeply inside `resolve_ident_somewhere`.
Now we explicitly pass a `ParentScope` to those functions instead, which includes the module and some other data describing our position in the crate relatively to which we resolve names.

Rustdoc was one of the users of `current_module`, it set it for resolving intra-doc links.
Now it passes it explicitly as an argument as well (I also supported resolving paths from rustdoc in unnamed blocks as a drive-by fix).

Visibility resolution is also changed to use early resolution (which is correct because it's used during the work of `BuildReducedGraphVisitor`, i.e. integration of a new AST fragment into the existing partially built module structures.) instead of untimely late resolution (which worked only due to restrictions on paths in visibilities like inability to refer to anything except ancestor modules).
This slightly regresses its diagnostics because late resolution has a more systematic error detection and recovery currently.
Due to changes in `current_module` and visibilities `BuildReducedGraphVisitor` ended up almost as heavily affected by this refactoring as late resolution.

Fixes https://github.com/rust-lang/rust/issues/63223 (due to visibility resolution changes).

### CODE REVIEW

**Comment by petrochenkov** [comment]
Actually, let's r? @matklad @Xanewok on this, nobody knows the nuances of this code anyway and this PR doesn't do many nuanced things beside perhaps some of the `current_module` changes.

@rust-lang/compiler @davidtwco are also welcome to review.

The PR is better reviewed in per-commit fashion.
Some of the commits are large because they move code between files and `impl`s, but they don't do much else.

File: src/librustc_resolve/build_reduced_graph.rs [hunk 1]
```diff
@@ -395,7 +400,7 @@ impl<'a> Resolver<'a> {
                 });
                 self.potentially_unused_imports.push(directive);
                 let imported_binding = self.import(binding, directive);
-                if ptr::eq(self.current_module, self.graph_root) {
+                if ptr::eq(parent, self.graph_root) {
```
Comment (eddyb) [inline src/librustc_resolve/build_reduced_graph.rs]: Random aside, is there any chance we could compare modules by `DefId` or something?
Also, the way they are used (for block scopes as well), they feel a bit more like "(def) scopes" than "modules" (but we also have the other kind of scope, the `Rib`s, which can contain runtime variables).

**Comment by eddyb** [comment]
@petrochenkov I'm a bit curious, what "inputs" does late resolution have, i.e. what does it use from the `Resolver`, and could that information be saved globally after expansion finishes?
I know early resolution and expansion are intertwined, but maybe we can make late resolution a query sooner than we otherwise could (without the early/late split).

File: src/librustc_resolve/late.rs [hunk 2]
```diff
@@ -66,6 +66,230 @@ impl PatternSource {
...
+            (PathSource::Type, false) => "E0412",
+            (PathSource::Struct, true) => "E0574",
+            (PathSource::Struct, false) => "E0422",
+            (PathSource::Expr(..), true) => "E0423",
+            (PathSource::Expr(..), false) => "E0425",
+            (PathSource::Pat, true) | (PathSource::TupleStruct, true) => "E0532",
+            (PathSource::Pat, false) | (PathSource::TupleStruct, false) => "E0531",
+            (PathSource::TraitItem(..), true) => "E0575",
+            (PathSource::TraitItem(..), false) => "E0576",
+        }
```
Comment (eddyb) [inline src/librustc_resolve/late.rs]: @nikomatsakis This sort of error code juggling feels very silly tbh.

File: src/librustc_resolve/build_reduced_graph.rs (hunk 1 above)
Comment (petrochenkov) [reply src/librustc_resolve/build_reduced_graph.rs]: `Module` in resolve is either a `mod`, or `enum`, or `trait`, or an `ast::Block`.
The naming seems ancient, I don't know why it was chosen, perhaps there was only `mod` initially, but then it got extended.

Modules and many other things (including those without IDs) are allocated on the resolver arena, so they use referential equality pretty regularly.
In this case `ptr::eq` is just simpler, I guess, than extracting the optional `DefId`s and comparing them if they exist.

Comment (petrochenkov) [reply src/librustc_resolve/build_reduced_graph.rs]: Oh, wait, two `Module`s can have the same `DefId`, see https://github.com/rust-lang/rust/pull/57181.
I think that case cannot apply to `self.graph_root` specifically, though.

**Comment by petrochenkov** [comment]
@eddyb 
>I'm a bit curious, what "inputs" does late resolution have, i.e. what does it use from the `Resolver`

Most of the resolver fields, I think?
Many of them for diagnostics only (e.g. if you want to suggest `println!(...)` instead of `println(...)` you need the macro-related tables).
The primary ones are probably `Module` and `NameBinding` data in `Resolver::arenas` + `Resolver::module_map` containing the module structure of the crate and names registered in those modules.

A few of the `Resolver` fields are pass-specific and can can be moved into pass-specific structures + some fields are "output only" (e.g. `label_res_map`), but by the most part the purpose of `Resolver` as a structure is to hoard data so it can then answer queries from late resolution, HIR lowering, or rustdoc.

**Comment by matklad** [comment]
I love this so much, thank you! Could you perhaps add a short comment at the top of `late.rs` describing what it is? I feel like I understand this from the issue description, but, if I were just reading the code, I would be left wondering what this thing is and why there's no `early.rs` :)

**Comment by matklad** [comment]
also cc @vlad20012: this new setup might serve as an inspiration for rewriting IntelliJ Rust name resolution :)

**Comment by petrochenkov** [comment]
Some file-level docs are added in https://github.com/rust-lang/rust/pull/63535.

**Pull request merged**